crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -7185,9 +7180,30 @@ impl<'db> TypeInferenceBuilder<'db> {
                 self.infer_type_expression(slice);
                 value_ty
             }
+            Type::ClassLiteral(ClassLiteralType::Generic(generic_class)) => {
+                let specialized_class =
+                    self.infer_explicit_class_specialization(subscript, value_ty, generic_class);
+                specialized_class
+                    .in_type_expression(self.db())
+                    .unwrap_or(Type::unknown())
+            }
+            Type::ClassLiteral(ClassLiteralType::NonGeneric(_)) => {
+                // TODO: Once we know that e.g. `list` is generic, emit a diagnostic if you try to
+                // specialize a non-generic class.
+                self.infer_type_expression(slice);
+                todo_type!("specialized non-generic class")
+            }
+            Type::StringLiteral(_) => {
+                // Don't emit a diagnostic, since we haven't determined the type of the deferred
+                // string annotation yet.
+                self.infer_type_expression(slice);
+                Type::unknown()
+            }
```

Comment:
Trying to understand what case this is handling -- did it come up in ecosystem checks? It seems like it would be permitting an annotation like `"list"[str]`, but AFAIK we don't need to support that, it isn't allowed. Pyright [doesn't allow it](https://pyright-play.net/?strict=true&enableExperimentalFeatures=true&code=CYUwZgBGAUAeBcEBEAbAlgZwC5INpoDssBdASggFoA%2BCQreAKAmYgCcQsBXVgiWXAAzEGDIA). Am I misunderstand what this is for?

Reply:
We have this in our mdtests, e.g. here

https://github.com/astral-sh/ruff/blob/b63de60399f77afb4d5e0032d60b59c186b70cfb/crates/red_knot_python_semantic/resources/mdtest/generics/classes.md?plain=1#L166-L168

Should we change that to `"C[T]"` per pyright's suggestion?

Reply:
Changed to `C[T]` and verified that the noisy new diagnostics didn't fire after removing the `StringLiteral` match arm.

Reply:
It looks like [mypy does support this](https://mypy-play.net/?mypy=latest&python=3.12&gist=da9a197f4b774d839a106518e1ab905d). But I don't feel like we should go out of our way to support it, IMO it's an odd thing to do, and the alternative (quote the entire annotation) is easy and the more obvious choice anyway. So I support what you've done here.